File: src-tauri/src/twitch/stream.rs
use std::collections::HashMap;

use anyhow::{anyhow, Result};
use serde::Serialize;

use crate::util;

use super::{
    main,
    query::{PlaybackAccessTokenQuery, PlaybackAccessTokenResponse, UseLiveQuery, UseLiveResponse},
};

const USHER_API: &str = "https://usher.ttvnw.net/api/channel/hls";

#[derive(Serialize, Debug)]
pub struct LiveNow {
    pub username: String,
    pub started_at: String,
}

pub async fn fetch_live_now(usernames: Vec<String>) -> Result<HashMap<String, LiveNow>> {
    if usernames.is_empty() {
        return Err(anyhow!("No usernames provided"));
    }

    let mut query: Vec<UseLiveQuery> = Vec::new();

    for username in usernames {
        if username.is_empty() {
            continue;
        }

        query.push(UseLiveQuery::new(&username));
    }

    let response: Vec<UseLiveResponse> = match main::send_query(query).await {
        Ok(data) => data,
        Err(err) => {
            return Err(anyhow!("Requesting UseLive: {err}"));
        }
    };

    let mut live_now: HashMap<String, LiveNow> = HashMap::new();

    for obj in response {
        if obj.data.user.stream.is_none() {
            continue;
        }

        let stream = obj.data.user.stream.unwrap();
        let username = obj.data.user.login;

        let live = LiveNow {
            username: username.clone(),
            started_at: stream.created_at,
        };

        live_now.insert(username, live);
    }

    Ok(live_now)
}

#[tauri::command]
pub async fn fetch_stream_playback(username: &str, backup: bool) -> Result<String, String> {
    if username.is_empty() {
        return Err(String::from("No username provided"));
    }

    let gql = PlaybackAccessTokenQuery::new(username, backup);

    let response: PlaybackAccessTokenResponse = match main::send_query(gql).await {
        Ok(response) => response,
        Err(err) => {
            return Err(format!("Requesting stream playback: {err}"));
        }
    };

    let stream_playback = response.data.stream_playback_access_token;

    let url = playlist_url(
        username,
        backup,
        &stream_playback.signature,
        &stream_playback.value,
    );

    Ok(url)
}

fn playlist_url(username: &str, backup: bool, signature: &str, token: &str) -> String {
    let mut url = format!("{USHER_API}/{username}.m3u8");

    let random_number = util::random_number(1_000_000, 10_000_000);

    if backup {
        url.push_str(&format!("?platform=ios&supported_codecs=h264&player=twitchweb&fast_bread=true&p={random_number}&sig={signature}&token={token}"));
    } else {
        url.push_str(&format!("?platform=web&supported_codecs=av1,h265,h264&allow_source=true&player=twitchweb&fast_bread=true&p={random_number}&sig={signature}&token={token}"));
    }

    url.to_string()
}

File: src/main/twitch/stream.ts
import {
	PlaybackAccessTokenQuery,
	PlaybackAccessTokenResponse,
	sendQuery,
	StreamInfoQuery,
	StreamInfoResponse,
	UseLiveQuery,
	UseLiveResponse
} from './query'

export async function fetchLiveNow(usernames: string[]): Promise<LiveNow[]> {
	if (usernames.length === 0) {
		return []
	}

	const query: UseLiveQuery[] = []

	for (const username of usernames) {
		if (username.length === 0) {
			continue
		}

		query.push(UseLiveQuery.new(username))
	}

	const response = await sendQuery<UseLiveResponse[]>(JSON.stringify(query))

	const liveNow: LiveNow[] = []

	for (const obj of response) {
		if (obj.data.user.stream === null) {
			continue
		}

		const stream = obj.data.user.stream
		const username = obj.data.user.login

		const live: LiveNow = {
			username: username,
			started_at: stream.createdAt
		}

		liveNow.push(live)
	}

	return liveNow
}

export async function fetchStream(username: string, backup: boolean): Promise<string> {
	const gql = [UseLiveQuery.new(username), PlaybackAccessTokenQuery.new(username, backup)]

	const response = await sendQuery<[UseLiveResponse, PlaybackAccessTokenResponse]>(
		JSON.stringify(gql)
	)

	const useLive = response[0]
	if (useLive.data.user.stream === null) {
		return ''
	}

	const streamPlayback = response[1].data.streamPlaybackAccessToken

	let url = `https://usher.ttvnw.net/api/channel/hls/${username}.m3u8`

	const randomNumber = Math.floor(Math.random() * 10_000_000) + 1_000_000

	if (backup) {
		url += `?platform=ios&supported_codecs=h264&player=twitchweb&fast_bread=true&p=${randomNumber}&sig=${streamPlayback.signature}&token=${streamPlayback.value}`
	} else {
		url += `?platform=web&supported_codecs=av1,h265,h264&allow_source=true&player=twitchweb&fast_bread=true&p=${randomNumber}&sig=${streamPlayback.signature}&token=${streamPlayback.value}`
	}

	return url
}

export async function fetchStreamInfo(username: string): Promise<StreamInfo> {
	const gql = StreamInfoQuery.new(username)

	const response = await sendQuery<StreamInfoResponse>(JSON.stringify(gql))

	if (!response.data.user) {
		throw new Error(`User '${username}' not found`)
	}

	if (!response.data.user.stream) {
		throw new Error(`User '${username}' is not live`)
	}

	const stream = response.data.user.stream

	let box_art = 'https://static-cdn.jtvnw.net/ttv-static/404_boxart-144x192.jpg'
	if (stream.game && stream.game.id) {
		const box_art_url = `https://static-cdn.jtvnw.net/ttv-boxart/${stream.game.id}-144x192.jpg`

		try {
			const response = await fetch(box_art_url)
			if (response.redirected) {
				box_art = `https://static-cdn.jtvnw.net/ttv-boxart/${stream.game.id}_IGDB-144x192.jpg`
			} else if (response.status === 200) {
				box_art = box_art_url
			}
		} catch (err) {
			console.error(`Error fetching box art: ${err}`)
		}
	}

	return {
		title: stream.title,
		game: stream.game ? stream.game.name : 'N/A',
		box_art: box_art,
		started_at: stream.createdAt,
		viewer_count: stream.viewersCount
	}
}
